src/coreclr/vm/excep.cpp

```diff
@@ -3226,196 +2858,326 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
 
 
 
-void StackTraceInfo::Init()
+#ifndef TARGET_UNIX // Watson is supported on Windows only
+void SetupWatsonBucket(UINT_PTR currentIP, CrawlFrame* pCf)
+{
+    Thread *pThread = GetThread();
+
+    if (pThread && (currentIP != 0))
+    {
+        // Setup the watson bucketing details for the initial throw
+        // callback only if we dont already have them.
+        ThreadExceptionState *pExState = pThread->GetExceptionState();
+        if (!pExState->GetFlags()->GotWatsonBucketDetails())
+        {
+            // Adjust the IP if necessary.
+            UINT_PTR adjustedIp = currentIP;
+            // This is a workaround copied from above.
+            if (!(pCf->HasFaulted() || pCf->IsIPadjusted()) && adjustedIp != 0)
+            {
+                adjustedIp -= 1;
+            }
+
+            // Setup the bucketing details for the initial throw
+            SetupInitialThrowBucketDetails(adjustedIp);
+        }
+    }
+}
+#endif // !TARGET_UNIX
+
+// Ensure that there is space for one more element in the stack trace array.
+void StackTraceInfo::EnsureStackTraceArray(StackTraceArray *pStackTrace, size_t neededSize)
 {
     CONTRACTL
     {
-        NOTHROW;
-        GC_NOTRIGGER;
-        MODE_ANY;
-        FORBID_FAULT;
+        GC_TRIGGERS;
+        THROWS;
+        PRECONDITION(CheckPointer(pStackTrace));
     }
     CONTRACTL_END;
 
-    LOG((LF_EH, LL_INFO10000, "StackTraceInfo::Init (%p)\n", this));
+    StackTraceArray newStackTrace;
+    GCPROTECT_BEGIN(newStackTrace);
 
-    m_pStackTrace = NULL;
-    m_cStackTrace = 0;
-    m_dFrameCount = 0;
-    m_cDynamicMethodItems = 0;
-    m_dCurrentDynamicIndex = 0;
+    size_t stackTraceCapacity = pStackTrace->Capacity();
+    if (neededSize > stackTraceCapacity)
+    {
+        S_SIZE_T newCapacity = S_SIZE_T(stackTraceCapacity) * S_SIZE_T(2);
+        if (newCapacity.IsOverflow() || (neededSize > newCapacity.Value()))
+        {
+            newCapacity = S_SIZE_T(neededSize);
+        }
+
+        stackTraceCapacity = newCapacity.Value();
+
+        // Allocate a new array with the needed size
+        newStackTrace.Allocate(stackTraceCapacity);
+        if (pStackTrace->Get() != NULL)
+        {
+            STRESS_LOG3(LF_EH, LL_INFO1000, "EnsureStackTraceArray copying from stackTrace=%p to stackTrace=%p, owner thread was %04x\n", OBJECTREFToObject(pStackTrace->Get()), OBJECTREFToObject(newStackTrace.Get()), pStackTrace->GetObjectThread()->GetOSThreadId());
+            // Copy the original array to the new one
+            newStackTrace.CopyDataFrom(*pStackTrace);
+            // TODO: cleanup this - make the method have delta as the parameter instead
```

Comment:
Fix TODO?

Reply:
I think that that line of code the comment refers to is stale from the time when I wasn't returning copies of the array from GetStackTrace for stack traces created by external threads. I'll double check.